src/server.cpp

```diff
@@ -12,107 +11,53 @@ std::string packetTypeToString(int type)
     }
     return "uncnown";
 }
 
-void Server::printPacket(const Packet& p)
-{
-    std::cout << "Packet type: " << packetTypeToString(p.type()) << "\n";
-
-    std::cout << "ID: " << std::to_string(p.id()) << "\n";
-
-    std::cout << "Attributes:\n";
-    for (const auto& ap : p.attributes())
-        std::cout << "\t" << m_dictionaries.attributes().name(ap->type()) << ": " << ap->toString() << "\n";
-
-    for (const auto& ap : p.vendorSpecific())
-    {
-        const std::string vendorName = m_dictionaries.vendorNames().name(ap->vendorId());
-        std::cout << "\tVendor-Specific" << ": " << vendorName << "\n";
-        std::cout << "\t" << m_dictionaries.vendorAttributes().name(vendorName, ap->vendorType()) << ": " << ap->toString() << "\n";
-    }
-}
-
+using Server = RadProto::Server;
 Server::Server(boost::asio::io_service& io_service, const std::string& secret)
-      : m_socket(io_service, udp::endpoint(udp::v4(), 9999)),
-        m_secret(secret),
-        m_dictionaries("/usr/share/freeradius/dictionary")
+    : m_socket(io_service, udp::endpoint(udp::v4(), 9999)),
+      m_secret(secret)
 {
-    startReceive();
 }
 
-void Server::startReceive()
+void Server::asyncReceive(const std::function<void(const error_code&, const std::optional<Packet>&)>& callback)
 {
     m_socket.async_receive_from(boost::asio::buffer(m_recvBuffer), m_remoteEndpoint,
-        [this](const error_code& error, std::size_t bytes) {handleReceive(error, bytes);});
+       [this, callback](const error_code& error, std::size_t bytes) {handleReceive(error, bytes, callback);});
 }
 
-void Server::handleReceive(const error_code& error, std::size_t bytes)
+void Server::asyncSend(const Packet& response, const std::function<void(const error_code&)>& callback)
 {
-    if (error)
-    {
-        std::cout << "Error async_receivw_from: " << error.message() << "\n";
-        return;
-    }
+    m_socket.async_send_to(boost::asio::buffer(response.makeSendBuffer(m_secret)), m_remoteEndpoint,
+       [this, callback](const error_code& ec, std::size_t /*bytesTransferred*/) {handleSend(ec, callback);});
+}
 
+void Server::handleReceive(const error_code& error, std::size_t bytes, std::function<void(const error_code&, const std::optional<Packet>&)> callback)
+{
     if (bytes < 20)
-    {
-        std::cout << "Error: request length less than 20 bytes\n";
-        return;
-    }
-
+        callback(Error::numberOfBytesIsLessThan20, std::nullopt);
     try
     {
-        Packet packet = makeResponse(Packet(m_recvBuffer, bytes, m_secret));
-
-        std::cout << "Response packet\n";
-        printPacket(packet);
-
-        m_socket.async_send_to(boost::asio::buffer(packet.makeSendBuffer(m_secret)), m_remoteEndpoint,
-           [this](const error_code& /*ec*/, std::size_t /*bytesTransferred*/) {handleSend();});
+        const Packet request = Packet(m_recvBuffer, bytes, m_secret);
+        callback(error, std::make_optional<Packet>(request));
```

Comment:
You do not need to create `Packet` first and put it into `std::make_optional`. This function works as a constructor, it accepts a template parameter (e.g. `Packet`) and constructs it inside using arguments. Consider the following two examples:
```c++
const Packet request = Packet(buffer, bytes, secret); // Constructs 'request' object using 'buffer', 'bytes' and 'secret'
const auto op1 = std::make_optional<Packet>(); // Constructs an object of type 'Packet' using 'request' as an argumet (calls copy constructor)

// ---***---

// Here is a better way
const auto op2 = std::make_optional<Packet>(buffer, bytes, secret); // Constructs an object of type 'Packet' using 'buffer', 'bytes' and 'secret'
```